### PR TITLE
Added Stream.bracketCase

### PR DESCRIPTION
**Pull request by mpilquist**
This needs some tests but I'm interested in feedback on the implementation. I basically just chased types.

### CODE REVIEW

**Comment by pchlupacek** [comment]
:-) This seems to enter shallow waters of resource management. Just to understand whats the benefit of adding this? (I am sure there is one, just need to quickly get on it)

**Comment by mpilquist** [comment]
@pchlupacek Read from this comment to the end: https://github.com/functional-streams-for-scala/fs2/issues/1140#issuecomment-400077235

**Comment by djspiewak** [comment]
This sort of thing would be profoundly useful in many cases. One example that we ran into in quasar is this general pattern:

1. Allocate some sort of bucket
2. Do work on completing that bucket
3. Bracket around **2** with a release which closes and publishes said bucket

In the even that the above is interrupted, step 3 happens, but what we *actually* want is to close and *clean up* the bucket, rather than publishing it. The way we're achieving this right now is with a `Signal` which we `set` as an action which is appended to **2** and then checked in **3**. This is quite clunky, and I think `bracketCase` would provide a more elegant answer.

**Comment by SystemFw** [comment]
I'd like to have a look at this before merging.
Definite :tada: for having it though

**Comment by pchlupacek** [comment]
@mpilquist @djspiewak , please note there is a significant difference between `interruption` and `cancellation`.  I am definitely in favor somehow to add this feature, I am just wondering between somehow clearly understand all the implications what this change will mean to interruption and how we can make sure the users won't somehow mismatch these two concepts that do completely different stuff :-) 

For example in case of what I understand from @djspiewak comment, this talks about `interruption` which is however completely unrelated to `cancellation`, and I didn't completely understand if the `interruption` of the stream causes somehow mimic the `cancellation` as you requested.

**Comment by pchlupacek** [comment]
Just probably to clarify, what I am pointing out to this PR is that `interrupting` a stream will allow to signal `ExitCase.Cancel` and as such signal/ (or eventually recover) from this.

File: core/shared/src/main/scala/fs2/internal/Resource.scala [hunk 1]
```diff
@@ -147,8 +147,10 @@ private[internal] object Resource {
                   else {
                     // scope is closed and this is last lease, assure finalizer is removed from the state and run
                     F.flatten(state.modify { s =>
-                      // previous finalizer shall be alwayy present at this point, this shall invoke it
-                      s.copy(finalizer = None) -> s.finalizer.getOrElse(F.pure(Right(())))
+                      // previous finalizer shall be always present at this point, this shall invoke it
+                      s.copy(finalizer = None) -> s.finalizer
+                        .map(_(ExitCase.Canceled))
```
Comment (AdamChlupacek) [inline core/shared/src/main/scala/fs2/internal/Resource.scala]: I think this should be `ExitCase.Completed` as releasing a lease seems like normal completion of the resource. It is just delayed.

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/internal/Resource.scala]: Thanks, good catch

**Comment by djspiewak** [comment]
> For example in case of what I understand from @djspiewak comment, this talks about interruption which is however completely unrelated to cancellation, and I didn't completely understand if the interruption of the stream causes somehow mimic the cancellation as you requested.

Ah, I could use some clarification on this point! What I'm talking about is stream interruption, for example with things like `interruptWhen`.

**Comment by AdamChlupacek** [comment]
Good job. This all looks good. I am just afraid that people will mistaken `cancelation` and `interruption`. As for interrupted streams this will pass the `ExitCase.Completed` to the `bracketCase`. Now I think we could easily make it that in case of interruption we could pass `ExitCase.Canceled` to the `bracketCase`, but I am not entirely sure if we want to make them one case without the possibility of saying which is which.

**Comment by AdamChlupacek** [comment]
@mpilquist oh, I wrote my comment too soon, seems like you already put the cases together.

**Comment by mpilquist** [comment]
@AdamChlupacek @djspiewak `ExitCase.Canceled` should work now for `stream.interruptWhen`. I'll work on some tests & docs tonight.

**Comment by SystemFw** [comment]
> Ah, I could use some clarification on this point! 

Same here actually, although it looks like things have been unified now regardless

**Comment by mpilquist** [comment]
Okay this should be good for review now. Tests are messy but par for the course for `ResourceSafetySpec`. I'd really like to redo them at some point but I'd rather we do that later than doing it incrementally on each PR.

**Comment by pchlupacek** [comment]
@mpilquist excellent. LGTM

**Pull request merged**